feat: check check constructor args type

fwx5618177 opened the pull request:
<!--
Thank you for using Hardhat and taking the time to send a Pull Request!

If you are introducing a new feature, please discuss it in an Issue or with someone from the team before submitting your change.

Please:
 - consider the checklist items below
 - keep the ones that make sense for your PR, and
 - DELETE the items that DON'T make sense for your PR.
-->

- [ ] Because this PR includes a **bug fix**, relevant tests have been included.
- [x] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [ ] I didn't do anything of this.

---

1. Added detection of parameters when passing arguments to the constructor. Currently, the detection is done on the backend, which makes it difficult to distinguish between strings and numbers in the tool. Therefore, a new feature has been submitted to address this issue.

This is the contract code:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.20;

contract MyContract {

    string public url;
    uint256 public add;

    constructor(string memory _url, uint256 _add) {
        url = _url;
        add = _add;
    }

    function setUrl(string memory _url) public {
        url = _url;
    }
}
```

Parameter passing file:
```js
module.exports = [
    123,
    123
];
```

Using the command:
```bash
npx hardhat verify --contract contracts/MyToken.sol:MyContract [ADDRESS] --constructor-args scripts/arguments.js
```

fwx5618177 commented:
@kanej  Maybe this pr not feature, it's fix?

schaable reviewed (approved):
This is a nice check when the constructor arguments are provided via a file or when the task is called programmatically. However, it doesn't have any effect when the arguments are passed through the CLI, as they are interpreted as strings. I've cleaned up the code, added a test, and included the changeset.

![image](https://github.com/NomicFoundation/hardhat/assets/1165424/5fd89428-38e1-471f-8815-eefac1129dd7)

fwx5618177 commented:
> This is a nice check when the constructor arguments are provided via a file or when the task is called programmatically. However, it doesn't have any effect when the arguments are passed through the CLI, as they are interpreted as strings. I've cleaned up the code, added a test, and included the changeset.
> 
> ![image](https://private-user-images.githubusercontent.com/1165424/330634167-5fd89428-38e1-471f-8815-eefac1129dd7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3MzkzNDMsIm5iZiI6MTcxNTczOTA0MywicGF0aCI6Ii8xMTY1NDI0LzMzMDYzNDE2Ny01ZmQ4OTQyOC0zOGUxLTQ3MWYtODgxNS1lZWZhYzExMjlkZDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTVUMDIxMDQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWM0ZGM0MTQ4YTUyNTQxNTViNzIzMTg3NzdiNTA5MTAxZDNmZDE3NzIzMTBiNDJlY2UwNmZjMTgzMzcxZWU0OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.WhXHvedWTvH2OZA6TrWuShiOyR4XF_txChVLmjCcAY8)

Got it. Maybe We need to transform string to a number and check if it's NaN or not.

schaable commented:
> Got it. Maybe We need to transform string to a number and check if it's NaN or not.

I don't think we need to, as the verification succeeds anyway.
